pennylane/transforms/decompositions/clifford_t/clifford_t_transform.py

```diff
@@ -0,0 +1,479 @@
+            # Check if the subsequent gate exists in the vicinity
+            copied_ops.pop(next_gate_idx)
+            next_gate_idx = find_next_gate(curr_gate.wires, copied_ops)
+            if next_gate_idx is None:
+                break
+            next_gate = copied_ops[next_gate_idx]
+
+        # Replace the current gate, add it to merged list and pop it from orginal one
+        merged_ops.append(curr_gate.__class__(*cumulative_angles, wires=curr_gate.wires))
+
+    return merged_ops, number_ops
+
+
+# pylint: disable= too-many-nested-blocks, too-many-branches, too-many-statements, unnecessary-lambda-assignment
+@transform
+def clifford_t_decomposition(
+    tape: QuantumTape,
+    epsilon=1e-4,
+    max_expansion=6,
+    method="sk",
+    **method_kwargs,
+) -> (Sequence[QuantumTape], Callable):
+    r"""Unrolls a circuit into the Clifford+T basis using the optimal ancilla-free approximation of :class:`~.RZ` operations.
+
+    This method first decomposes the gate operations to a basis comprised of Clifford, :class:`~.T`, :class:`~.RZ` and
+    :class:`~.GlobalPhase` operations (and their adjoints). The Clifford gates include the following PennyLane operations:
+
+    - Single qubit gates - :class:`~.Identity`, :class:`~.PauliX`, :class:`~.PauliY`, :class:`~.PauliZ`,
+      :class:`~.SX`, :class:`~.S`, and :class:`~.Hadamard`.
+    - Two qubit gates - :class:`~.CNOT`, :class:`~.CY`, :class:`~.CZ`, :class:`~.SWAP`, and :class:`~.ISWAP`.
+
+    Then the leftover single qubit :class:`~.RZ` operations are approximated in the Clifford+T basis with
+    :math:`\epsilon > 0` error using the Solovay-Kitaev algorithm described in
+    `Dawson and Nielsen (2005) <https://arxiv.org/abs/quant-ph/0505030>`_.
+
+    Args:
+        tape (QuantumTape): A quantum tape
+        epsilon (float): The maximum permissible error. Defaults to ``0.0001``.
+        max_expansion (int): The depth to be used for tape expansion before manual decomposition to Clifford+T basis is applied.
+        method (str): Method to be used for Clifford+T decomposition. Default value is ``"sk"`` for Solovay-Kitaev
+        **method_kwargs: Keyword argument to pass options for the ``method`` used for decompositions.
+
+    Keyword Args:
+        Options (*):
+
+            * **max_depth** (int), **basis_set** (list[str]), **basis_length** (int):
+              arguments for the ``"sk"`` method, i.e., for performing Solovay-Kitaev decomposition using :func:`~.sk_decomposition`
+
+    Returns:
+        pennylane.QNode or qfunc or tuple[List[.QuantumTape], function]: If a QNode is passed,
+        it returns a QNode with the transform added to its transform program. If a tape is passed,
+        returns a tuple containing a list of quantum tapes to be evaluated, and a function to be
+        applied to these tape executions.
+
+    Raises:
+        ValueError: If any gate operation does not have a rule for its decomposition
```

Comment:
```suggestion
        ValueError: If a gate operation does not have a decomposition when required
```
Mainly for my understanding - could this work on gates that only have a matrix method defined?